### ISSUE TITLE
[BUG]: Stale CUDA error causes CUB calls to fail

### ISSUE THREAD

**Issue by dkolsen-pgi** [NVIDIA/cccl#1791]
### Is this a duplicate?

- [X] I confirmed there appear to be no [duplicate issues](https://github.com/NVIDIA/cccl/issues) for this bug and that I agree to the [Code of Conduct](CODE_OF_CONDUCT.md)

### Type of Bug

Runtime Error

### Component

CUB

### Describe the bug

PR #1641 causes CUB calls to fail if `cudaGetLastError()` returns something other than `cudaSuccess`.

If the last CUDA call before a call to CUB algorithm (or a Thrust algorithm that is implemented with CUB) failed and the CUDA error is not cleared, the CUB algorithm call will fail as if the uncleared error caused the failure.

(NVHPC tracking: FS#35618)

### How to Reproduce

```cpp
#include <thrust/device_vector.h>
#include <thrust/host_vector.h>
#include <thrust/reduce.h>
#include <stdio.h>

int main() {
  thrust::host_vector<int> hv;
  for (int i = 0; i < 1000; ++i) {
    hv.push_back(i);
  }
  thrust::device_vector<int> dv{hv};
  cudaDeviceProp properties;
  cudaGetDeviceProperties(&properties, 100); // invalid device number
                                             // ignore the error
  int sum = thrust::reduce(dv.begin(), dv.end());
  printf("sum = %d\n", sum);
}
```
When compiled with CUDA 12.5, which has a CCCL release from before #1641, the program runs successfully.
```
$ /proj/cuda/12.5/Linux_x86_64/bin/nvcc -arch=native -x cu  stale-error.cpp
$ ./a.out
sum = 499500
```
But when NVCC is redirected to use CCCL head, the program fails at runtime.
```
$ /proj/cuda/12.5/Linux_x86_64/bin/nvcc -arch=native -x cu -I/proj/cuda/cccl/main/libcudacxx/include -I/proj/cuda/cccl/main/thrust -I/proj/cuda/cccl/main/cub  stale-error.cpp
$ ./a.out
terminate called after throwing an instance of 'thrust::THRUST_200500_800_NS::system::system_error'
  what():  after reduction step 1: cudaErrorInvalidDevice: invalid device ordinal
Aborted
```

If the device index of `100` is changed to `0`, the call to `cudaGetDeviceProperties` succeeds and the entire program runs successfully with the latest CCCL.

### Expected behavior

CUB should not get tripped up by an uncleared CUDA error.

**Comment by jrhemstad** [comment]
> CUB should not get tripped up by an uncleared CUDA error.

I don't think this is possible in general unless we synchronized on entry to every CUB call. 

At any time during the execution of a CUB call, a prior asynchronous CUDA error may get raised and caught somewhere inside the CUB call. 

Even if we limited the scope to errors already pending before the CUB call, our only option would be to use `cudaGetLastError()` to clear the unhandled error state. But what should CUB do if there was a pending error? Throw? Ignore it? 

If we throw, then we're right back to the situation that started this issue in the first place.

If we ignore it, then we took it upon ourselves to clear the global error state on the users behalf. That isn't really CUB's responsibility and can hide errors from users.

In other words, there's no way for us to ignore incoming errors without clearing the global error state. Clearing the global error state on behalf of users isn't a good idea.

Long story short, I think the only answer here is that calling a CUB API with a pending CUDA error state is UB.

**Comment by dkolsen-pgi** [comment]
> I think the only answer here is that calling a CUB API with a pending CUDA error state is UB.

Ugh.  Okay.  I don't have the expertise to push back against that.  I'll have to throw this back to the application owner and tell them "don't do that."

This is the second time recently that I have been burned by the lousy error handling API in CUDA.
